File: src/common/py/core/messaging/handlers/message_handler.py
import typing
from dataclasses import dataclass

from ..message import MessageType
from .message_context import MessageContextType

MessageHandler = typing.Callable[[MessageType, MessageContextType], None]


@dataclass(frozen=True)
class MessageHandlerMapping:
    """
    Mapping from a message name filter to a message handler.
    
    Attributes:
        filter: The message name filter.
        handler: The message handler.
        message_type: The message type the handler expects.
        is_async: Whether the handler should be invoked asynchronously in its own thread.
    """
    filter: str
    handler: MessageHandler
    message_type: type[MessageType]
    is_async: bool = False
    
    def __str__(self) -> str:
        return f"{self.filter} -> {str(self.handler)} [{str(self.message_type)}]"


MessageHandlerMappings = typing.List[MessageHandlerMapping]

File: src/common/web/core/messaging/handlers/MessageHandler.ts
import { type Constructable } from "../../../utils/Types";
import { Message } from "../Message";
import { MessageContext } from "./MessageContext";

export type MessageHandler = (msg: Message, ctx: MessageContext) => void;

/**
 * Mapping from a message name filter to a message handler.
 */
export class MessageHandlerMapping {
    /**
     * @param filter - The message name filter.
     * @param handler - The message handler.
     * @param messageType - The message type the handler expects.
     */
    public constructor(readonly filter: string, readonly handler: MessageHandler, readonly messageType: Constructable) {
    }

    /**
     * Gets the string representation of the handler mapping.
     */
    public toString(): string {
        return `${this.filter} -> ${String(this.handler)} [${String(this.messageType)}]`;
    }
}

export type MessageHandlerMappings = MessageHandlerMapping[];
